lib/shared/open_ai/api_client.dart

```diff
@@ -0,0 +1,369 @@
+import '../../db/meal_plan_entry.dart';
+import '../../db/recipe_ingredient.dart';
+import '../../db/user.dart';
+import '../../db/storage_ingredient.dart';
+import '../../db/meal_plan.dart';
+import '../../db/recipe.dart';
+import '../../db/recipe_step.dart';
+import 'package:isar/isar.dart';
+
+class ApiClient {
+  static bool _validateRequest(
+      List<dynamic> items, User user, String itemType) {
+    if (items.isEmpty) {
+      developer.log('No $itemType provided', name: 'OpenAI');
+      return false;
+    }
+    if (user.apiKey == null) {
+      developer.log('No api key provided', name: 'OpenAI');
+      return false;
+    }
+    return true;
+  }
+
+  static Future<Ingredient> _findOrCreateIngredient(
+    Map<String, dynamic> ingredientData,
+    List<Ingredient> existingIngredients,
+    Isar isar,
+  ) async {
+    Ingredient ingredient = existingIngredients.firstWhere(
+        (i) =>
+            i.name == ingredientData['name'] &&
+            i.unit == ingredientData['unit'],
+        orElse: () => Ingredient()
+          ..name = ingredientData['name']
+          ..unit = ingredientData['unit']);
+
+    if (ingredient.id == Isar.autoIncrement) {
+      await isar.ingredients.put(ingredient);
+    }
+    return ingredient;
+  }
+
+  /// Analyzes images to detect ingredients and their quantities, storing them in the database.
+  ///
+  /// [images] List of base64 encoded image strings to analyze
+  /// [user] The user making the request, must have a valid API key
+  /// [isar] Database instance for storing the results
+  ///
+  /// Returns a list of [StorageIngredient] objects if successful, null otherwise.
+  static Future<List<StorageIngredient>?> generateStorageIngredients(
+      List<String> images, User user, Isar isar) async {
+    if (!_validateRequest(images, user, 'images')) return null;
+
+    final requestBody = RequestBody(
+      model: 'gpt-4o',
+      messages: [
+        Message(
+          role: 'user',
+          content: [
+            Content(
+              type: 'text',
+              value:
+                  'Analyze these images and identify all ingredients with their approximate quantities.',
+            ),
+            ...images.map((image) => Content(
+                  type: 'image_url',
+                  value: {'url': "data:image/png;base64,$image"},
+                )),
+          ],
+        ),
+      ],
+      functions: [
+        AiFunction(
+          name: 'detect_ingredients',
+          description: 'Detects ingredients from images',
+          parameters: {
+            'type': 'object',
+            'properties': {
+              'ingredients': {
+                'type': 'array',
+                'items': {
+                  'type': 'object',
+                  'properties': {
+                    'name': {
+                      'type': 'string',
+                      'description': 'Name of the ingredient'
+                    },
+                    'quantity': {
```

Comment:
In dem StorageIngredient haben wir es immer count genannt. Sollte das vielleicht einheitlich sein?